File: hono-bun/src/schemas/attachment-schemas.ts
import { z } from "@hono/zod-openapi";

import { PaginationMetaSchema, PaginationQuerySchema } from "@/core/pagination";

export const AttachmentSchema = z.object({
  id: z.string().uuid().openapi({ description: "Attachment ID" }),
  taskId: z.string().uuid().openapi({ description: "Parent task ID" }),
  fileName: z
    .string()
    .openapi({ description: "Original file name", example: "report.pdf" }),
  fileKey: z.string().openapi({ description: "S3 object key" }),
  mimeType: z
    .string()
    .openapi({ description: "MIME type", example: "application/pdf" }),
  fileSize: z
    .number()
    .int()
    .openapi({ description: "File size in bytes", example: 1024 }),
  uploadedBy: z
    .string()
    .uuid()
    .nullable()
    .openapi({ description: "Uploader user ID" }),
  createdAt: z.string().datetime().openapi({ description: "Upload timestamp" }),
});

export const AttachmentCreateSchema = z.object({
  fileName: z.string().min(1).max(255).openapi({
    description: "Original file name",
    example: "report.pdf",
  }),
  mimeType: z.string().min(1).max(128).openapi({
    description: "MIME type",
    example: "application/pdf",
  }),
  fileSize: z.number().int().min(1).openapi({
    description: "File size in bytes",
    example: 1024,
  }),
});

export const AttachmentListQuerySchema = PaginationQuerySchema;

export const AttachmentListResponseSchema = z.object({
  data: z.array(AttachmentSchema),
  meta: PaginationMetaSchema,
});

export const AttachmentUploadResponseSchema = z.object({
  uploadUrl: z.string().url().openapi({
    description: "Presigned URL for uploading the file",
  }),
  attachment: AttachmentSchema,
});

File: fastapi/src/schemas/attachment_schemas.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from src.core.pagination import PaginationMeta, PaginationQuery


class AttachmentResponse(BaseModel):
    id: UUID = Field(description="Attachment ID")
    task_id: UUID = Field(description="Parent task ID")
    file_name: str = Field(description="Original file name", examples=["report.pdf"])
    file_key: str = Field(description="S3 object key")
    mime_type: str = Field(description="MIME type", examples=["application/pdf"])
    file_size: int = Field(description="File size in bytes", examples=[1024])
    uploaded_by: UUID | None = Field(default=None, description="Uploader user ID")
    created_at: datetime = Field(description="Upload timestamp")


class AttachmentCreate(BaseModel):
    file_name: str = Field(min_length=1, max_length=255, description="Original file name", examples=["report.pdf"])
    mime_type: str = Field(min_length=1, max_length=128, description="MIME type", examples=["application/pdf"])
    file_size: int = Field(gt=0, description="File size in bytes", examples=[1024])


class AttachmentListQuery(PaginationQuery):
    pass


class AttachmentListResponse(BaseModel):
    data: list[AttachmentResponse]
    meta: PaginationMeta


class AttachmentUploadResponse(BaseModel):
    upload_url: str = Field(description="Presigned URL for uploading the file")
    attachment: AttachmentResponse
